tools/torch/data.lua

```diff
@@ -41,33 +41,205 @@ local function all_keys(cursor_,key_,op_)
+        im_hsv[2].image.saturate(im_hsv[2]) -- bound saturation between 0 and 1
+    end
+    if augHSV.V >0.001 then
+        local rng_v = torch.normal(0, augHSV.V)
+        im_hsv[3] = torch.add(im_hsv[3], torch.FloatTensor(1,im_hsv:size(2),im_hsv:size(3)):fill(rng_v))
+        im_hsv[3].image.saturate(im_hsv[3]) -- bound value between 0 and 1
+    end
+    return image.hsv2rgb(im_hsv)
+end
+
+-- Scale and Rotation augmentation (warping)
+-- Parameters:
+-- @param im (tensor): input image
+-- @param augRot (float): extremes of random rotation, uniformly distributed between 
+-- @param augScale (float): the standard deviation of the extra scaling factor
+local warp = function(im, augRot, augScale)
+    -- A nice function of scale is 0.05 (stddev of scale change), 
+    -- and a nice value for ration is a few degrees or more if your dataset allows for it
+
+    local width = im:size()[3] 
+    local height = im:size()[2]
+
+    -- Scale <0=zoom in(+rand crop), >0=zoom out
+    local scale_x = 0
+    local scale_y = 0
+    local move_x = 0
+    local move_y = 0
+    if augScale > 0.001 then
+        scale_x = torch.normal(0, augScale) -- normal distribution
+        -- Given a zoom in or out, we move around our canvas.
+        scale_y = scale_x -- keep aspect ratio the same
+        move_x = torch.uniform(-scale_x, scale_x)
+        move_y = torch.uniform(-scale_y, scale_y)
+    end
+
+    -- Angle of rotation
+    local rot_angle = (torch.rand(1)[1]-0.5)* 2 * augRot -- (degrees) uniform distribution [-augRot : augRot)
+
+    -- x/y grids
+    local grid_x = torch.ger( torch.ones(height), torch.linspace(-1-scale_x,1+scale_x,width) )
+    local grid_y = torch.ger( torch.linspace(-1-scale_y,1+scale_y,height), torch.ones(width) )
+
+    local flow = torch.FloatTensor()
+    flow:resize(2,height,width)
+    flow:zero()
+
+    -- Apply scale
+    flow_scale = torch.FloatTensor()
+    flow_scale:resize(2,height,width)
+    flow_scale[1] = grid_y
+    flow_scale[2] = grid_x
+    flow_scale[1]:add(1+move_y):mul(0.5) -- move ~[-1 1] to ~[0 1]
+    flow_scale[2]:add(1+move_x):mul(0.5) -- move ~[-1 1] to ~[0 1]
+    flow_scale[1]:mul(height-1)
+    flow_scale[2]:mul(width-1)
+    flow:add(flow_scale)
+
+    if augRot > 0.001 then
+        -- Apply rotation through rotation matrix
+        local flow_rot = torch.FloatTensor()
+        flow_rot:resize(2,height,width)
+        flow_rot[1] = grid_y * ((height-1)/2) * -1
+        flow_rot[2] = grid_x * ((width-1)/2) * -1
+        view = flow_rot:reshape(2,height*width)
+        function rmat(deg)
+          local r = deg/180*math.pi
+          return torch.FloatTensor{{math.cos(r), -math.sin(r)}, {math.sin(r), math.cos(r)}}
+        end
+
+        local rotmat = rmat(rot_angle)
+        local flow_rotr = torch.mm(rotmat, view)
+        flow_rot = flow_rot - flow_rotr:reshape( 2, height, width )
+        flow:add(flow_rot)
+    end
+
+    return image.warp(im, flow, 'bilinear', false)
```

Comment:
did you choose to use `image.warp` because it does scaling and rotation in a single pass so it's more efficient that using `image.scale` and `image.rotate`?